File: NSIDC/MOD10C1.A2005018.006.2016141204712.hdf.py
"""
Copyright (C) 2014 John Evans

This example code illustrates how to access and visualize a NSIDC
MOD10C1 HDF-EOS2 Grid file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python MOD10C1.A2005018.006.2016141204712.hdf.py

The HDF file must be in your current working directory.

Tested under: Python 2.7.15 :: Anaconda custom (64-bit)
Last updated: 2019-02-21
"""

import os
import re

import matplotlib as mpl
import matplotlib.pyplot as plt
from mpl_toolkits.basemap import Basemap
import pyproj
import numpy as np

USE_GDAL = False

def run(FILE_NAME):
    
    # Identify the data field.
    DATAFIELD_NAME = 'Day_CMG_Snow_Cover'

    if USE_GDAL:    
        import gdal
        GRID_NAME = 'MOD_CMG_Snow_5km'    
        gname = 'HDF4_EOS:EOS_GRID:"{0}":{1}:{2}'.format(FILE_NAME,
                                                         GRID_NAME,
                                                         DATAFIELD_NAME)
        gdset = gdal.Open(gname)
        data = gdset.ReadAsArray()
        
        # Read projection parameters.
        x0, xinc, _, y0, _, yinc = gdset.GetGeoTransform()
        nx, ny = (gdset.RasterXSize, gdset.RasterYSize)

        del gdset

    else:
        from pyhdf.SD import SD, SDC
        hdf = SD(FILE_NAME, SDC.READ)

        # Read dataset.
        data2D = hdf.select(DATAFIELD_NAME)
        data = data2D[:,:].astype(np.float64)

        # Read global attribute.
        fattrs = hdf.attributes(full=1)
        ga = fattrs["StructMetadata.0"]
        gridmeta = ga[0]
            
        # Read projection parameters. 
        # The needed information is in a global attribute called 'StructMetadata.0'.  
        # Use regular expressions to tease out the extents of the grid. 
        ul_regex = re.compile(r'''UpperLeftPointMtrs=\(
                                  (?P<upper_left_x>[+-]?\d+\.\d+)
                                  ,
                                  (?P<upper_left_y>[+-]?\d+\.\d+)
                                  \)''', re.VERBOSE)
        match = ul_regex.search(gridmeta)
        x0 = np.float(match.group('upper_left_x')) / 1e6
        y0 = np.float(match.group('upper_left_y')) / 1e6

        lr_regex = re.compile(r'''LowerRightMtrs=\(
                                  (?P<lower_right_x>[+-]?\d+\.\d+)
                                  ,
                                  (?P<lower_right_y>[+-]?\d+\.\d+)
                                  \)''', re.VERBOSE)
        match = lr_regex.search(gridmeta)
        x1 = np.float(match.group('lower_right_x')) / 1e6
        y1 = np.float(match.group('lower_right_y')) / 1e6
        ny, nx = data.shape
        xinc = (x1 - x0) / nx
        yinc = (y1 - y0) / ny
        
    # Construct the grid.  It's already in lat/lon.
    x = np.linspace(x0, x0 + xinc*nx, nx)
    y = np.linspace(y0, y0 + yinc*ny, ny)
    lon, lat = np.meshgrid(x, y)

    m = Basemap(projection='cyl', resolution='l',
                llcrnrlat=-90, urcrnrlat = 90,
                llcrnrlon=-180, urcrnrlon = 180)
    m.drawcoastlines(linewidth=0.5)
    m.drawparallels(np.arange(-90., 120., 30.), labels=[1, 0, 0, 0])
    m.drawmeridians(np.arange(-180, 180., 45.), labels=[0, 0, 0, 1])

    # Bin the data as follows:
    # 0% snow
    # 1-99% snow
    # 100% snow
    # lake ice (107)
    # night (111)
    # inland water (237)
    # ocean (239)
    # cloud-obscured water (250)
    # data not mapped (253)
    # fill (255)
    lst = ['#00ff00', # 0% snow
           '#888888', # 1-99% snow
           '#ffffff', # 100% snow
           '#ffafff', # lake ice
           '#000000', # night
           '#0000cc', # inland water
           '#0000dd', # ocean
           '#63c6ff', # cloud-obscured water
           '#00ffcc', # data not mapped
           '#8928dd'] # fill
    cmap = mpl.colors.ListedColormap(lst)
    bounds = [0, 1, 100, 107, 111, 237, 239, 250, 253, 255, 256]
    norm = mpl.colors.BoundaryNorm(bounds, cmap.N)
    
    # Render the image in the projected coordinate system.
    m.pcolormesh(lon[::2,::2], lat[::2,::2], data[::2,::2],
                 latlon=True, cmap=cmap, norm=norm)

    long_name = 'Day CMG Snow Cover'
    basename = os.path.basename(FILE_NAME)
    plt.title('{0}\n{1}'.format(basename, long_name))
    fig = plt.gcf()

    color_bar = plt.colorbar(orientation='horizontal')
    color_bar.set_ticks([0.5, 50, 103, 109, 180, 237.8, 242, 251.5, 254.5, 255.5])
    color_bar.set_ticklabels(['0%\nsnow', '1-99%\nsnow', '100%\nsnow', 'lake\nice',
                              'night', 'inland\nwater', 'ocean',
                              'cloud\n-obscured\nwater', 'data\nnot\nmapped',
                              'fill'])
    color_bar.draw_all()

    pngfile = "{0}.py.png".format(basename)
    fig.savefig(pngfile)


if __name__ == "__main__":

    hdffile = 'MOD10C1.A2005018.006.2016141204712.hdf'
    run(hdffile)
    

File: NSIDC/MOD10C1_A2005018_006_2016141204712_hdf.m
%
%  This example code illustrates how to access and visualize
% NSIDC MOD10C1 L3 HDF-EOS2 Grid file in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example,
% please use the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA
%  HDF/HDF-EOS data product that is not listed in the HDF-EOS
%  Comprehensive Examples page (http://hdfeos.org/zoo), 
% feel free to contact us at eoshelp@hdfgroup.org or post it at the 
% HDF-EOS Forum (http://hdfeos.org/forums).
%                 
% Usage: save this script and run (without .m at the end)
% 
% $matlab -nosplash -nodesktop -r MOD10C1_A2005018_006_2016141204712_hdf
%
% Tested under: MATLAB R2018b
% Last updated: 2019-02-22

import matlab.io.hdf4.*
import matlab.io.hdfeos.*
                 
% Define file name, grid name, and data field.
FILE_NAME='MOD10C1.A2005018.006.2016141204712.hdf';
GRID_NAME='MOD_CMG_Snow_5km';
DATAFIELD_NAME='Day_CMG_Snow_Cover';

% Open the HDF-EOS2 Grid file.
file_id = gd.open(FILE_NAME, 'rdonly');

% Read data from a data field.
grid_id = gd.attach(file_id, GRID_NAME);
[data, lat, lon] = gd.readField(grid_id, DATAFIELD_NAME, [], [], []);

% Detach from the Grid Object.
gd.detach(grid_id);

% Close the file.
gd.close(file_id);

% Get information about the spatial extents of the grid.
% [xdimsize, ydimsize, upleft, lowright, status] = hdfgd('gridinfo', grid_id);

% Detach from the Grid object.
% hdfgd('detach', grid_id);

% Close the file.
% hdfgd('close', file_id);

% Convert the data to double type for plot.
data=double(data);

% Read attributes from the data field.
SD_id = sd.start(FILE_NAME, 'rdonly');
sds_index = sd.nameToIndex(SD_id, DATAFIELD_NAME);
sds_id = sd.select(SD_id, sds_index);

% Read units from the data field.
units_index = sd.findAttr(sds_id, 'units');
units = sd.readAttr(sds_id, units_index);

% Read long_name from the data field.
long_name_index = sd.findAttr(sds_id, 'long_name');
long_name = sd.readAttr(sds_id, long_name_index);

% Terminate access to the corresponding data set.
sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);


% The following will return unique key values used in the dataset.
z = unique(data);

% To get a similar image to NSIDC browse image [1], 
% construct a color table based on the following assignment:
%
% Key      R  G   B    Name
% ==========================
%  0%     0   100 0    dark green
%  1-99%  127 127 127  grey
%  100%   255 255 255  white
%  107    255 176 255  pink  
%  111    0   0   0    black
%  237    0   0   255  blue
%  239    0   0   205  medium blue                 
%  250    100 200 255
%  253    255 0   255  magenta
%  255    138 42  226  blue violet
%
%
%  We added two more (0% and 1-99%) entries for ice coverage to get better
%  image.


% Plot the data using surfacem and axesm.
latlim=[floor(min(min(lat))),ceil(max(max(lat)))];
lonlim=[floor(min(min(lon))),ceil(max(max(lon)))];

f = figure('Name', FILE_NAME, ...
           'Renderer', 'zbuffer', ...
           'Position', [0,0,800,600], ...
           'visible','off');

axesm('MapProjection','eqdcylin','Frame','on','Grid','on', ...
      'MapLatLimit',latlim,'MapLonLimit',lonlim, ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel','south')
coast = load('coast.mat');

% Here is the color map 
cmap=[[0.00 1.00 0.00];  ... %   0% dark green
      [0.50 0.50 0.50];  ... %   1-99% grey
      [1.00 1.00 1.00];  ... %   100% white
      [1.00 0.69 1.00];  ... %   107 pink 
      [0.00 0.00 0.00];  ... %   111 black
      [0.00 0.00 1.00];  ... %   237 blue                     
      [0.00 0.00 0.80];  ... %   239
      [0.39 0.78 1.00];  ... %   250
      [0.00 1.00 0.78];  ... %   253                      
      [0.54 0.16 0.87]]; ... %   255
colormap(cmap);

% Put 1%-99% data under one grey label.
data((data > 0) & (data < 100)) = 99;

% Construct a discrete data for plot.
z = unique(data);
k = size(z);

% Create an array for tick label.
y = zeros(k, 'double');

% There are k different boxes in the colorbar 
% and the value starts from 1 to m.
% Thus, we should increment by (k-1)/k to position
% labels properly starting form ((k-1)/k)/2.
x = 1 + ((k-1)/k)/2;

for m = 1:k
    y(m) = x; 
    data(data == z(m)) = double(m);
    x = x+(k-1)/k;    
end

surfm(lat,lon,data);

caxis([1 m]);
h = colorbar('YTickLabel',...
         {'0% snow', '1-99% snow', '100% snow', 'lake ice', 'night', ...
          'inland water', 'ocean', ...    
          'cloud obsc. water', 'data not mapped', 'fill'}, 'YTick', y);

plotm(coast.lat,coast.long,'k')
                 
tightmap;
                 
title({FILE_NAME;...
       long_name}, ...
      'FontSize',12,'FontWeight','bold', 'Interpreter', 'none');

% Save image.
set (get(h, 'title'), 'string', units, 'FontSize',12,'FontWeight','bold');
saveas(f, [FILE_NAME '.m.png']);
exit;


